src/sys/unix/selector/poll.rs

```diff
@@ -0,0 +1,725 @@
+        let mut pending_removal = self.pending_removal.lock().unwrap();
+        if let Some(idx) = pending_removal.iter().position(|&pending| pending == fd) {
+            pending_removal.remove(idx);
+        }
+        drop(pending_removal);
+
+        self.modify_fds(|fds| {
+            if fds.fd_data.contains_key(&fd) {
+                return Err(io::Error::new(
+                    io::ErrorKind::AlreadyExists,
+                    "\
+                    same file descriptor registered twice for polling \
+                    (an old file descriptor might have been closed without deregistration)\
+                    ",
+                ));
+            }
+
+            let poll_fds_index = fds.poll_fds.len();
+            let record = Arc::new(RegistrationRecord::new());
+            fds.fd_data.insert(
+                fd,
+                FdData {
+                    poll_fds_index,
+                    token,
+                    shared_record: record.clone(),
+                },
+            );
+
+            fds.poll_fds.push(PollFd(libc::pollfd {
+                fd,
+                events: interests_to_poll(interests),
+                revents: 0,
+            }));
+
+            Ok(record)
+        })
+    }
+
+    pub fn reregister(&self, fd: RawFd, token: Token, interests: Interest) -> io::Result<()> {
+        self.modify_fds(|fds| {
+            let data = fds.fd_data.get_mut(&fd).ok_or(io::ErrorKind::NotFound)?;
+            data.token = token;
+            let poll_fds_index = data.poll_fds_index;
+            fds.poll_fds[poll_fds_index].0.events = interests_to_poll(interests);
+
+            Ok(())
+        })
+    }
+
+    pub fn deregister(&self, fd: RawFd) -> io::Result<()> {
+        self.deregister_all(&[fd])
+            .map_err(|_| io::ErrorKind::NotFound)?;
+        Ok(())
+    }
+
+    /// Perform a modification on `fds`, interrupting the current caller of `wait` if it's running.
+    fn modify_fds<T>(&self, f: impl FnOnce(&mut Fds) -> T) -> T {
+        self.waiting_operations.fetch_add(1, Ordering::SeqCst);
+
+        // Wake up the current caller of `wait` if there is one.
+        let sent_notification = self.notify_waker.wake().is_ok();
+
+        let mut fds = self.fds.lock().unwrap();
+
+        // If there was no caller of `wait` our notification was not removed from the pipe.
+        if sent_notification {
+            self.notify_waker.ack_and_reset();
+        }
+
+        let res = f(&mut *fds);
+
+        if self.waiting_operations.fetch_sub(1, Ordering::SeqCst) == 1 {
+            self.operations_complete.notify_one();
+        }
+
+        res
+    }
+
+    /// Special optimized version of [Self::deregister] which handles multiple removals
+    /// at once.  Ok result if all removals were performed, Err if any entries
+    /// were not found.
+    fn deregister_all(&self, targets: &[RawFd]) -> Result<(), ()> {
```

Comment:
I don't think this function, as used by `SelectorState::select`, should return an error, it should just ignore them and continue removing all the fds. Only the single `deregister` should return an error.

Reply:
It returns so that deregister() can reuse the code and forward the error from it

Reply:
I understand that, but on line 416 it does the following:

```rust
            for target in targets {
                let data = fds.fd_data.remove(&target).ok_or(())?;
```
So that the first error it hits will be returned, not removing the other fds. For a single `deregister` call by the user we want this, but not in `Selector::select`, here we want to continue to remove the other fds as well (we ignore any errors from `deregister_all` any way).

Reply:
Oh good catch and I misunderstood what you meant by return in that sense.  Yeah, I'll fix it.